### ISSUE TITLE
Bug: [no-restricted-imports] Potentially invalid rule JSON schema (`required: undefined`).

### ISSUE THREAD

**Issue by virtuallyunknown** [typescript-eslint/typescript-eslint#10493]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

Greetings. I am importing the typescript-eslint rules and storing information about them such as name, meta, schemas and others in a database, so later on I can compare changes between different versions programmatically. Recently I noticed a potential issue with `no-restricted-imports`.

It appears that one of the keys in it's schema is of type `undefined`, which to my knowledge is not valid JSON. Here is the issue in question:

```js
import { plugin } from 'typescript-eslint';
const rule = plugin.rules['no-restricted-imports'];

console.log(
    rule.meta.schema.anyOf[1].items[0].properties.patterns.anyOf[1].items
);
```
```js
// console output:
{
  type: 'object',
  additionalProperties: false,
  properties: {
    // skipped for brevity
  },
  required: undefined
  //        ^^^^^^^^^
}
```

I believe the issue is caused by the following lines of code in the library:

https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/rules/no-restricted-imports.ts#L103

https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/rules/no-restricted-imports.ts#L111

https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/rules/no-restricted-imports.ts#L138

Attempting to `JSON.stringify` the schema will drop the key `required` key altogether, so there is no way for me to store it as it is in a JSON file, though I imagine it wouldn't be a trivial task even if I was using something else.

If this is intentional and not a bug, then I apologize in advance, though I checked all other rule schemas and I couldn't find other places where `undefined` values were present.

### Reproduction Repository Link

https://github.com/virtuallyunknown/typescript-eslint-no-restricted-imports-schema-undefined

### Repro Steps

1) clone the repo
2) run `node index.js`
3) it will show the contents of `schema.anyOf[1].items[0].properties.patterns.anyOf[1].items` and it's `required` key which is of type `undefined`.

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `typescript-eslint`     | `8.18.0`     | `8.18.0` |
| `@typescript-eslint/eslint-plugin`     | `8.18.0` |
| `node`                                 | `23.3.0` |

**Comment by bradzacher** [comment]
This is dependent on your underlying version of ESLint -- the weird code exists to allow us to support the rule's schema as it has changed over the ESLint versions.

You might want your analyser to be a little smarter than just `JSON.stringify`ing so you can capture such things if you think they are important to your work.

Note that the exact structure of rule schemas is not part of our public API and so we make no guarantees about consistency or structure between releases and may make changes in non-major releases.

**Comment by virtuallyunknown** [comment]
Hey, I appreciate your quick reply.

> the exact structure of rule schemas is not part of our public API and so we make no guarantees about consistency or structure between releases and may make changes in non-major releases

I anticipate schema changes between releases, that's totally fine. The issue I'm facing is that it's not valid as JSON anymore because of the `undefined` value, and therefor I can't store the schema in it's original form. As a result of that, each time I do a comparison between the schema in my JSON database and the schema imported from `typescript-eslint`, a change is detected because of the missing key that gets dropped.

From all the rules in `typescript-eslint` this is the only one with a schema which contains an `undefined` as a value, and this is most definitely beyond my understanding, but wouldn't it be better to use `null` here?

In any case, if things remain unchanged here, could you possibly give me an advice on how to proceed? The options I have as I see are:

1. Do nothing for the time being and just wait for a future ESLint update which might make the problem go away.
2. JSON.stringify and then parse all schemas before comparing them to the ones in my own database.
3. Write a custom transformer that handles `undefined` values and converts it into `null` or maybe use a 3rd party library like superjson.

Just to give a little context - the purpose of the tool I made is to create/export eslint configuration files, and I validate these configurations against the rule schemas I store in the database.

Cheers.
